pkg/archive/spec.go

```diff
@@ -1,15 +1,171 @@
 package archive
 
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+var (
+	alphanumeric = regexp.MustCompile(`[^a-zA-Z0-9]+`)
+
+	ExtensionMap = map[string]string{
+		"bookworm": "deb",
+		"bullseye": "deb",
+		"buster":   "deb",
+		"focal":    "deb",
+		"jammy":    "deb",
+		"rhel9":    "rpm",
+		"rhel8":    "rpm",
+		"centos7":  "rpm",
+		"mariner2": "rpm",
+		"windows":  "zip",
+	}
+
+	OSMap = map[string]string{
+		"bookworm": "debian",
+		"bullseye": "debian",
+		"buster":   "debian",
+		"focal":    "ubuntu",
+		"jammy":    "ubuntu",
+		"rhel9":    "el9",
+		"rhel8":    "el8",
+		"centos7":  "el7",
+		"mariner2": "cm2",
+		"windows":  "windows",
+	}
+
+	VersionMap = map[string]string{
+		"bookworm": "12",
+		"bullseye": "11",
+		"buster":   "10",
+		"focal":    "20.04",
+		"jammy":    "22.04",
+		"rhel9":    "el9",
+		"rhel8":    "el8",
+		"centos7":  "el7",
+		"mariner2": "cm2",
+	}
+)
+
 type Spec struct {
 	Pkg      string `json:"package"`
 	Distro   string `json:"distro"`
 	Arch     string `json:"arch"`
-	Repo     string `json:"repo"`
-	Commit   string `json:"commit"`
+	Repo     string `json:"repo" hash:"ignore"`
+	Commit   string `json:"commit" hash:"ignore"`
 	Tag      string `json:"tag"`
 	Revision string `json:"revision"`
 }
 
+func (spec *Spec) StoragePath() (string, error) {
+	pkg := spec.Pkg
+	pkgOS := spec.OS()
+	version := fmt.Sprintf("%s+azure", spec.Tag)
+	distro := spec.Distro
+	sanitizedArch := strings.ReplaceAll(spec.Arch, "/", "_")
+
+	base, err := spec.Basename()
+	if err != nil {
+		return "", err
+	}
+
+	storagePath := fmt.Sprintf("%s/%s/%s/%s_%s/%s", pkg, version, distro, pkgOS, sanitizedArch, base)
+
+	return storagePath, nil
+}
+
+// This logic is arbitrary, but the output must be reproducible. This is used
+// to generate filenames for artifacts.
+func (s *Spec) Hash() (string, error) {
+	v := reflect.ValueOf(s)
```

Comment:
Seems like we don't really need to use reflect here since we already know all the fields to include.

Reply:
When you're tired of working on pipelines, your brain will do some funny things ;)

But the advantage of keeping `reflect` here is that if the struct changes, the method doesn't need to be changed. Thoughts?

Reply:
My thought here is if you do need to add a field then you also need to determine up front if it should be ignored for hashing, so why not deal with that directly rather than through reflect?
I also think it is more clear what's going on in that case.

The flip side is right now adding a new field that should not be hashed but does get hashed is more likely (maybe?) to be noticed than the other way around.

Reply:
Done